flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java

```diff
@@ -0,0 +1,158 @@
+        extends IteratorSourceReaderBase<E, O, IterT, SplitT> {
+
+    private final GeneratorFunction<E, O> generatorFunction;
+
+    private BooleanSupplier couldExit;
+    private int snapshotsCompleted;
+    private int snapshotsToWaitFor = Integer.MAX_VALUE;
+    private boolean done;
+
+    public SourceReaderWithCheckpointsLatch(
+            SourceReaderContext context,
+            GeneratorFunction<E, O> generatorFunction,
+            @Nullable BooleanSupplier couldExit) {
+        super(context);
+        this.generatorFunction = checkNotNull(generatorFunction);
+        this.couldExit = couldExit;
+    }
+
+    public SourceReaderWithCheckpointsLatch(
+            SourceReaderContext context, GeneratorFunction<E, O> generatorFunction) {
+        super(context);
+        this.generatorFunction = checkNotNull(generatorFunction);
+    }
+
+    // ------------------------------------------------------------------------
+
+    @Override
+    public void start(SourceReaderContext context) {
+        try {
+            generatorFunction.open(context);
+        } catch (Exception e) {
+            throw new FlinkRuntimeException("Failed to open the GeneratorFunction", e);
+        }
+    }
+
+    @Override
+    public InputStatus pollNext(ReaderOutput<O> output) {
+        // This is the termination path after the test data has been emitted twice
+        if (done) {
+            if (couldExit != null) { // Termination is controlled externally
+                return couldExit.getAsBoolean()
+                        ? InputStatus.END_OF_INPUT
+                        : InputStatus.NOTHING_AVAILABLE;
+            } else {
+                return InputStatus.END_OF_INPUT;
+            }
+        }
+        // This is the initial path
+        if (currentSplit == null) {
+            InputStatus inputStatus = tryMoveToNextSplit();
+            switch (inputStatus) {
+                case MORE_AVAILABLE:
+                    emitElements(output);
+                    break;
+                case END_OF_INPUT:
+                    // This can happen if the source parallelism is larger than the number of
+                    // available splits
+                    return inputStatus;
+            }
+        } else {
+            // This is the path that emits the same split the second time
+            emitElements(output);
+            done = true;
+        }
+        availability = new CompletableFuture<>();
+        return InputStatus.NOTHING_AVAILABLE;
+    }
+
+    private void emitElements(ReaderOutput<O> output) {
+        iterator = currentSplit.getIterator();
+        while (iterator.hasNext()) {
+            E next = iterator.next();
+            O converted = convert(next);
+            output.collect(converted);
+        }
+        // Always wait for two snapshots after emitting the elements
+        snapshotsToWaitFor = 2;
+        snapshotsCompleted = 0;
```

Comment:
Why two snapshots?

Reply:
The whole endeavor is required to have a drop-in replacement for https://github.com/apache/flink/blob/release-1.18/flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/FiniteTestSource.java#L33-L35 without modifying the logic of existing tests that rely on it. This is how FiniteTestSource works.